src/tools/rest-backend/authentication_application.cpp

```diff
@@ -0,0 +1,412 @@
+		}
+		if (!std::regex_match (username, regex_username))
+		{
+			RootApp::setBadRequest (response (),
+						"The given username is not valid. The username must be "
+						"between 3 and 20 signs long and may contain only "
+						"letters a-z (lower- and upper-case), the numbers 0-9 "
+						"and dots (.) as well as dashes (-).",
+						"AUTH_REGISTER_INVALID_USERNAME");
+			return; // quit early
+		}
+		if (!std::regex_match (email, regex_email))
+		{
+			RootApp::setBadRequest (response (), "The given email is not a valid email.", "AUTH_REGISTER_INVALID_EMAIL");
+			return; // quit early
+		}
+		if (service::StorageEngine::instance ().userExists (username))
+		{
+			RootApp::setBadRequest (response (),
+						"An user with the given username does already exist. "
+						"Please choose another username.",
+						"AUTH_REGISTER_USERNAME_DOES_ALREADY_EXIST");
+			return; // quit early
+		}
+
+		// create user model
+		model::User u (username);
+		u.setPasswordHash (sha256::hash_hex (password));
+		u.setEmail (email);
+		u.setRank (1);
+		u.setCreatedAt (static_cast<long>(time (0)));
+
+		// store user
+		try
+		{
+			if (service::StorageEngine::instance ().createUser (u))
+			{
+				RootApp::setOk (response (), "Your account has been created successfully!",
+						"AUTH_REGISTER_USER_CREATED_SUCCESSFULLY");
+				return;
+			}
+			else
+			{
+				RootApp::setInternalServerError (response (),
+								 "Something went wrong while creating your account. "
+								 "Please try again.",
+								 "AUTH_REGISTER_UNKNOWN_ERROR");
+				return;
+			}
+		}
+		catch (exception::UserAlreadyExistsException & e)
+		{
+			RootApp::setBadRequest (response (),
+						"An user with the given username does already exist. "
+						"Please choose another username.",
+						"AUTH_REGISTER_USERNAME_DOES_ALREADY_EXIST");
+			return;
+		}
+	}
+	else if (request ().request_method () == "OPTIONS")
+	{
+		RootApp::setOk (response ());
+		return;
+	}
+	else
+	{
+		RootApp::setMethodNotAllowed (response ());
+		return;
+	}
+}
+
+void AuthenticationApp::user ()
+{
+	if (request ().request_method () == "GET")
+	{
+		try
+		{
+			model::User u = this->getCurrentUser (request ());
+			cppcms::json::value data;
+			data["status"] = static_cast<int>(cppcms::http::response::ok);
+			data["username"] = u.getUsername ();
+			data["rank"] = u.getRank ();
+			RootApp::setOk (response (), data);
+		}
+		catch (exception::NoCurrentUserException & e)
+		{
+			RootApp::setUnauthorized (response (), "You need to be authenticated to perform this action.",
+						  "NEED_AUTHENTICATION");
+		}
+		catch (exception::UserNotFoundException & e)
+		{
+			RootApp::setBadRequest (response (), "An error occured while processing your request.", "INTERNAL_ERROR_OCCURED");
+		}
+	}
+	else if (request ().request_method () == "OPTIONS")
+	{
+		RootApp::setOk (response ());
+		return;
+	}
+	else
+	{
+		RootApp::setMethodNotAllowed (response ());
+		return;
+	}
+}
+
+void AuthenticationApp::setrank ()
```

Comment:
I'm planning to add a very, very basic role-system with users, moderators and admins. So if this platform will ever be used, there is at least a way to edit and/or remove bad content. So this basically allows to set a rank.